cookbook/controller/argument_value_resolver.rst

```diff
@@ -0,0 +1,208 @@
+.. index::
+   single: Controller; Argument Value Resolvers
+
+Extending Action Argument Resolving
+===================================
+
+.. versionadded:: 3.1
+    The ``ArgumentResolver`` and value resolvers are added in Symfony 3.1.
+
+In the book, you've learned that you can get the :class:`Symfony\\Component\\HttpFoundation\\Request`
+object by adding a ``Request`` argument to your controller. This is done
+via the :class:`Symfony\\Component\\HttpKernel\\Controller\\ArgumentResolver`.
+By creating and registering custom argument value resolvers, you can extend
+this functionality.
+
+Functionality Shipped With The HttpKernel
+-----------------------------------------
+
+Symfony ships with four value resolvers in the HttpKernel:
+
+:class:`Symfony\\Component\\HttpKernel\\Controller\\ArgumentValueResolver\\ArgumentFromAttributeResolver`
+    Attempts to find a request attribute that matches the name of the argument.
+
+:class:`Symfony\\Component\\HttpKernel\\Controller\\ArgumentValueResolver\\RequestValueResolver`
+    Injects the current ``Request`` if type-hinted with ``Request``, or a
+    sub-class thereof.
+
+:class:`Symfony\\Component\\HttpKernel\\Controller\\ArgumentValueResolver\\DefaultValueResolver`
+    Will set the default value of the argument if present and the argument
+    is optional.
+
+:class:`Symfony\\Component\\HttpKernel\\Controller\\ArgumentValueResolver\\VariadicValueResolver`
+    Verifies in the request if your data is an array and will add all of
+    them to the argument list. When the action is called, the last (variadic)
+    argument will contain all the values of this array.
+
+.. note::
+
+    Prior to Symfony 3.1, this logic was resolved within the ``ControllerResolver``.
+    The old functionality is rewritten to the aforementioned value resolvers.
+
+Adding a Custom Value Resolver
+------------------------------
+
+Adding a new value resolver requires one class and one service defintion.
+In the next example, you'll create a value resolver to inject the ``User``
+object from the security system. Given you write the following action::
+
+    namespace AppBundle\Controller;
+
+    use AppBundle\Entity\User;
+    use Symfony\Component\HttpFoundation\Response;
+
+    class UserController
+    {
+        public function indexAction(User $user)
```

Comment:
I understand correctly comments below, you should add "= null" in the signature, right ?